pkg/csi/service/common/authmanager.go

```diff
@@ -135,44 +184,244 @@ func GenerateDatastoreIgnoreMapForBlockVolumes(ctx context.Context, vc *vsphere.
 		for dsURL, dsInfo := range dsURLTodsInfoMap {
 			dsURLs = append(dsURLs, dsURL)
 			dsInfos = append(dsInfos, dsInfo)
+			dsMoRef := dsInfo.Reference()
+			entities = append(entities, dsMoRef)
+		}
+	}
+
+	dsURLToInfoMap, err := getDatastoresWithRequiredPriv(ctx, vc, dsURLs, dsInfos, entities)
+	if err != nil {
+		log.Errorf("failed to get datastores with required priv. Error: %+v", err)
+		return nil, err
+	}
+	return dsURLToInfoMap, nil
+}
+
+// GenerateDatastoreMapForFileVolumes scans all datastores in Vcenter and do privilege check on those datastoes
+// It will return datastores which has the privileges for creating file volume
+func GenerateDatastoreMapForFileVolumes(ctx context.Context, vc *vsphere.VirtualCenter) (map[string]*cnsvsphere.DatastoreInfo, error) {
+	log := logger.GetLogger(ctx)
+	// get all vSAN datastores from VC
+	vsanDsURLToInfoMap, err := vc.GetVsanDatastores(ctx)
+	if err != nil {
+		log.Errorf("failed to get vSAN datastores with error %+v", err)
+		return nil, err
+	}
+	var allvsanDatastoreUrls []string
+	for dsURL := range vsanDsURLToInfoMap {
+		allvsanDatastoreUrls = append(allvsanDatastoreUrls, dsURL)
+	}
+	fsEnabledMap, err := IsFileServiceEnabled(ctx, allvsanDatastoreUrls, vc)
+	if err != nil {
+		log.Errorf("failed to get if file service is enabled on vsan datastores with error %+v", err)
+		return nil, err
+	}
+
+	var dsURLs []string
+	var dsInfos []*cnsvsphere.DatastoreInfo
+	var entities []types.ManagedObjectReference
+	dsURLToInfoMap := make(map[string]*cnsvsphere.DatastoreInfo)
+	for dsURL, dsInfo := range vsanDsURLToInfoMap {
+		if val, ok := fsEnabledMap[dsURL]; ok {
+			if val {
+				// go through all datastores to build entities
+				dsMoRef := dsInfo.Reference()
+				entities = append(entities, dsMoRef)
+				// Get all vSAN datastore URLs and Info which has vSAN FS enabled
+				dsURLs = append(dsURLs, dsURL)
+				dsInfos = append(dsInfos, dsInfo)
+			}
 		}
 	}
+	if len(entities) == 0 {
+		msg := "No file service enabled vsan datastore with required privileges for CSI user is present in the environment."
+		log.Infof(msg)
+		return dsURLToInfoMap, nil
+	}
+	dsURLToInfoMap, err = getDatastoresWithRequiredPriv(ctx, vc, dsURLs, dsInfos, entities)
```

Comment:
We already did the priv checking in `IsFileServiceEnabled`. Why is this required again?

Reply:
In `IsFileServiceEnabled` we only checked for clusters which have `Host.Config.Storage` permission on vSAN clusters with vSAN FS enabled. Then we get vSAN datastores from these clusters. 


But in `getDatastoresWithRequiredPriv` we want to check `Datastore.FileManagement` and `System.Read` permission on these vSAN datastores. Only those datastores would be picked for file volume provisioning.

Reply:
Why? CNS requires only `Host.Config.Storage` priv for file volume provisioning.

Reply:
True.